### ISSUE TITLE
Why is the mesh refined for tagging strategies even if the initial adaptive refinement is set to 0?

### ISSUE THREAD

**Issue by anne-glerum** [geodynamics/aspect#4507]
When ` Initial adaptive refinement = 0` and `Time steps between mesh refinement = 0`, the mesh will still be adaptively refined at t0 when the `Strategy` is one that tags cells, like `minimum refinement function`, given that the `Minimum refinement level` allows it. See examples below. This is not what I would expect, and I wonder whether it is intentional. 

Found when looking into #4088.

For these settings:
```
subsection Mesh refinement
  set Initial global refinement = 2
  set Initial adaptive refinement = 2 
  set Time steps between mesh refinement = 0 
  set Strategy = minimum refinement function, maximum refinement function
  set Minimum refinement level = 0
end
```
the mesh looks like this at t0:
![IGR2_IAR2](https://user-images.githubusercontent.com/7631624/155988988-5ce729b9-fac0-4afb-964b-e4b5773afdeb.png)

And this is what the log file says about the DOFs:
```
Number of degrees of freedom: 663,706 (93,206+46,128+93,206+11,739+46,603+46,603+46,603+46,603+46,603+46,603+46,603+46,603+46,603)
Number of mesh deformation degrees of freedom: 23478
Number of degrees of freedom: 1,721,180 (241,692+119,866+241,692+30,316+120,846+120,846+120,846+120,846+120,846+120,846+120,846+120,846+120,846)
Number of mesh deformation degrees of freedom: 60632
Number of degrees of freedom: 4,563,217 (640,636+318,859+640,636+80,224+320,318+320,318+320,318+320,318+320,318+320,318+320,318+320,318+320,318)
Number of mesh deformation degrees of freedom: 160448
```

For these settings:
```
subsection Mesh refinement
  set Initial global refinement = 4
  set Initial adaptive refinement  = 0 
  set Time steps between mesh refinement = 0 
  set Strategy = minimum refinement function, maximum refinement function
  set Minimum refinement level = 0
end
```
the t0 mesh looks like:
![IGR4_IAR0_MRL0_Stratminmax](https://user-images.githubusercontent.com/7631624/155989063-f776224e-8599-49a6-9aed-e3df889c1d3f.png)

and the log file states:
```
Number of degrees of freedom: 4,765,459 (669,078+332,660+669,078+83,792+334,539+334,539+334,539+334,539+334,539+334,539+334,539+334,539+334,539)
Number of mesh deformation degrees of freedom: 167584
```
 
For these settings:
```
subsection Mesh refinement
  set Skip solvers on initial refinement = false #same mesh for true
  set Initial global refinement = 4
  set Initial adaptive refinement = 0 
  set Time steps between mesh refinement = 0 
  #set Strategy = minimum refinement function, maximum refinement function
  set Minimum refinement level = 4 #same mesh for 0
end
```
the mesh has a global uniform resolution of level 4. `thermal energy density` is the default `Strategy`. 

log file:
```
Number of degrees of freedom: 19,732,815 (2,769,602+1,383,601+2,769,602+346,801+1,384,801+1,384,801+1,384,801+1,384,801+1,384,801+1,384,801+1,384,801+1,384,801+1,384,801)
Number of mesh deformation degrees of freedom: 693602
```

**Comment by anne-glerum** [comment]
I'll create a small test that reproduces the potential issue.

**Comment by anne-glerum** [comment]
```
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.4.0-pre (melt_visco_plastic_PM01, e87edc117)
--     . using deal.II 10.0.0-pre (master, d944f3d291)
--     .       with 32 bit indices and vectorization level 3 (512 bits)
--     . using Trilinos 12.18.1 
--     . using p4est 2.2.0
--     . running in OPTIMIZED mode
--     . running with 96 MPI processes
-----------------------------------------------------------------------------
```

**Comment by bangerth** [comment]
If I understand right what you say (a small test case would indeed help!) the user is asking for (i) no adaptive refinement, but (ii) also sets minimum and maximum refinement levels that are not satisfied by the global refinement steps. That's inconsistent, and because you set the interval between mesh refinement to zero, can never be made consistent. The question is what we should do with these conflicting instructions.

What is your suggestion of what the behavior *should* be?

**Comment by anne-glerum** [comment]
I've attached test cases 1-3, and figures for the corresponding meshes. 

What I would expect is:
1) No initial mesh refinement when `Initial adaptive refinement = 0`. 
2) The min/max minimum refinement strategy to act the same as other, non-tagging, strategies. So when `Initial adaptive refinement = 0`, neither leads to refinement at t0.


At the moment, when `Initial adaptive refinement = 0` and `Strategy = minimum refinement function, maximum refinement function`, the mesh is refined at t0.
But when `Initial adaptive refinement = 0` and `Strategy = thermal energy density`, the mesh is not refined.

[1.txt](https://github.com/geodynamics/aspect/files/8169474/1.txt)
![mesh_1](https://user-images.githubusercontent.com/7631624/156359560-a3934dc1-00ec-4645-a8dd-0c2deb15aa70.png)
[2.txt](https://github.com/geodynamics/aspect/files/8169475/2.txt)
![mesh_2](https://user-images.githubusercontent.com/7631624/156359583-f92fdfee-6ede-4eda-8d95-9bcfe19072ec.png)
[3.txt](https://github.com/geodynamics/aspect/files/8169476/3.txt)

![mesh_3](https://user-images.githubusercontent.com/7631624/156359612-1f09f39f-072b-4d74-a9cb-0d9e78013319.png)

**Comment by jdannberg** [comment]
I think I found the part of the code where this happens: In core.cc in `Simulator<dim>::run` we do

```
        // Instead of calling global_refine(n) we flag all cells for
        // refinement and then allow the mesh refinement plugins to unflag
        // the cells if desired. This procedure is repeated n times. If there
        // is no plugin that modifies the flags, it is equivalent to
        // refine_global(n).
        for (unsigned int n=0; n<parameters.initial_global_refinement; ++n)
          {
            for (const auto &cell : triangulation.active_cell_iterators())
              cell->set_refine_flag ();

            mesh_refinement_manager.tag_additional_cells ();
            triangulation.execute_coarsening_and_refinement();
            if (MappingQCache<dim> *map = dynamic_cast<MappingQCache<dim>*>(&(*mapping)))
              map->initialize(MappingQGeneric<dim>(4), triangulation);
          }
```
So it's not like the mesh is coarsened, or like it does any adaptive refinements. It's just that it doesn't do all of the global refinements if one of the criteria that tags cells removes the refinement flag. The `MaximumRefinementFunction<dim>::tag_additional_cells` clears the refinement flag whenever the refinement level is larger than what it allows. So I think what causes the behavior is not the minimum refinement function, it is the maximum refinement function. And it looks like this was implemented in the code intentionally (it would be very easy to just remove the `mesh_refinement_manager.tag_additional_cells ();` line in the code above, and that should change the behavior). 

So basically the question is: If you tell the code to refine for times, but then set the maximum refinement function to something that contains areas with a refinement level of less than 4, what should happen? I tend to think the current behavior makes sense, but I admit it is counterintuitive that you set 4 global refinements, but the model is not refined everywhere. Maybe we need to document this better for the "global refinement" parameter?

**Comment by anne-glerum** [comment]
That makes sense, the refine flags for the global refinements are removed based on the tagging strategy maximum refinement. 

I think it is counter-intuitive that it happens for global refinements though, especially as the mesh it produces is not as clean as the mesh produced by the same strategies but with adaptive refinement, see the first 2 figures of the first post.

The current description of `Initial global refinement` is 
```
      prm.declare_entry ("Initial global refinement", "2",
                         Patterns::Integer (0),
                         "The number of global refinement steps performed on "
                         "the initial coarse mesh, before the problem is first "
                         "solved there.");
``` 
What about
      prm.declare_entry ("Initial global refinement", "2",
                         Patterns::Integer (0),
                         "The number of global refinement steps performed on "
                         "the initial coarse mesh, before the problem is first "
                         "solved there. Note that tagging mesh refinement and "
                         "coarsening strategies such as the Maximum refinement "
                         "function can remove refinement flags in each refinement "
                         "step, such that the resulting mesh is not necessarily uniform. ");

**Comment by jdannberg** [comment]
Yes, I think that new description makes sense. 

But you are right, your mesh still looks strange in the second figure (the one with only global refinements). I could reproduce that in the example you posted, and it looks like there are cells that have a lower refinement level than what the maximum refinement function would allow for. I noticed that I get different meshes depending on how many processors I use. With 1 processor, I get the correct mesh even with 4 global and 0 adaptive refinements. 

I made a PR here #4513, can you check if that fixes it for you?

I also noticed that your maximum refinement function changes in levels directly at the boundary between cells in the mesh (in the example file you attached). For me, this led to the cells bordering the lines of y=0.25 and y=0.75 sometimes being refined, and sometimes not. 

If I changed the refinement functions to

```
  subsection Maximum refinement function
    set Coordinate system   = cartesian
    set Variable names      = x,y 
    set Function expression = if( y>=0.751, 4, if (y>=0.251, 2, 0)) 
  end
```
it looked much better.

**Comment by bangerth** [comment]
@anne-glerum Yes, I think the description makes sense to me as well. I would suggest adding a half-sentence somewhere to say that it is possible to provide conflicting information (e.g., refine globally 4 times, but only allow refinement level 2 with the max refinement level), and then say that if that is the case, whatever you say in the updated description is how that is interpreted. Want to make that into a patch?

@jdannberg The problem with the different behavior at the interface is that we look at vertices, not cell centers:
```
    template <int dim>
    void
    MinimumRefinementFunction<dim>::tag_additional_cells () const
    {
      for (const auto &cell : this->get_triangulation().active_cell_iterators())
        {
          if (cell->is_locally_owned())
            {
              bool refine = false;
              bool clear_coarsen = false;

              for (const unsigned int v : cell->vertex_indices())
                {
                  const Point<dim> vertex = cell->vertex(v);
                  const Utilities::NaturalCoordinate<dim> point =
                    this->get_geometry_model().cartesian_to_other_coordinates(vertex, coordinate_system);

                  const double minimum_refinement_level = min_refinement_level.value(Utilities::convert_array_to_point<dim>(point.get_coordinates()));

                  if (cell->level() <= rint(minimum_refinement_level))
                    clear_coarsen = true;
                  if (cell->level() <  rint(minimum_refinement_level))
                    {
                      refine = true;
                      break;
                    }
                }

              if (clear_coarsen)
                cell->clear_coarsen_flag ();
              if (refine)
                cell->set_refine_flag ();
            }
        }
    }
```
I think it would be reasonable to change this to only evaluate the cell center. This would also avoid the ambiguity about what to do if the criterion doesn't agree for all vertices of a cell.

What do you think? If you agree, I can prepare a patch.

**Comment by bangerth** [comment]
I opened #4521 for the min/max refinement level function. Looking forward to the patch updating the parameter description!

**Comment by gassmoeller** [comment]
Can this issue be closed now that #4521 and #4522 are merged?

**Comment by anne-glerum** [comment]
Yes, I'll close this issue. 

Related #4513.
